xarray/core/common.py

```diff
@@ -482,6 +482,54 @@ def where(self, cond):
         """
         return self._where(cond)
 
+    def sel_where(self, mask):
+        """Return an object of reduced shape for entries along each dimension where cond is
+        True. Other entries are excluded.
+
+        This operation follows the normal broadcasting and alignment rules that
+        xarray uses for binary arithmetic.
+
+        Parameters
+        ----------
+        cond : boolean DataArray or Dataset
+
+        Returns
+        -------
+        same type as caller with dimensions reduced for dim element where mask is True
+
+        Examples
+        --------
+
+        >>> import numpy as np
+        >>> a = xr.DataArray(np.arange(25).reshape(5, 5), dims=('x', 'y'))
+        >>> a.sel_where(np.logical_and(a > 6, a < 18))
+        <xarray.DataArray (x: 5, y: 5)>
+        array([[ nan,  nan,   7.,   8.,   9.],
+               [ 10.,  11.,  12.,  13.,  14.],
+               [ 15.,  16.,  17.,  nan,  nan],
+        Coordinates:
+          * x        (x) int64 1 2 3
+          * y        (y) int64 0 1 2 3 4
+        """
+        from .dataarray import DataArray
+        from .dataset import Dataset
+        # get mask with the minimal size needed for the Dataset
+        if isinstance(mask, Dataset):
+            clipmask = mask.to_array().any('variable')
+        elif isinstance(mask, DataArray):
+            clipmask = mask
+        elif isinstance(mask, bool):
+            return self.where(mask)
+        else:
+            raise ValueError("Mask argument is %r but must be a %r or %r" %
+                             (mask, Dataset, DataArray))
+
+        # clip the data corresponding to coordinate dims that are not used
+        clip = dict(zip(clipmask.dims,[np.unique(adim) for adim in np.where(clipmask)]))
+        self = self.sel(**clip)
+        mask = mask.sel(**clip)
```

Comment:
`np.where`/`np.nonzero` returns integers, so `clip` always has integer keys. Thus I think you need `isel` instead of `sel`. I would probably use `isel` on mask, then `sel` on self with indexes from mask.